File: 엄마호랑이반/10주차 문제/최혜령/과제진행하기.py
def solution(plans):
    # 시간을 분 단위로 변환하는 함수
    def to_minutes(time_str):
        h, m = map(int, time_str.split(":"))
        return h * 60 + m
    
    # 데이터 전처리: 시작 시간을 분으로 변환하고 소요시간도 정수로 변환
    tasks = []
    for name, start, duration in plans:
        tasks.append([name, to_minutes(start), int(duration)])
    
    # 시작 시간 기준으로 정렬
    tasks.sort(key=lambda x: x[1])
    
    stack = []  # 중단된 과제를 저장할 스택
    answer = []  # 완료된 과제 순서
    
    for i in range(len(tasks)-1):
        current_task = tasks[i]
        next_task = tasks[i+1]
        
        current_name, current_start, current_duration = current_task
        next_start = next_task[1]
        
        # 현재 과제 완료 시각
        current_end = current_start + current_duration
        
        # 다음 과제 시작 전에 현재 과제를 완료할 수 있는 경우
        if current_end <= next_start:
            answer.append(current_name)
            
            # 남는 시간 동안 이전에 중단된 과제 처리
            extra_time = next_start - current_end
            while extra_time > 0 and stack:
                suspended_name, remaining_time = stack.pop()
                if remaining_time <= extra_time:
                    answer.append(suspended_name)
                    extra_time -= remaining_time
                else:
                    stack.append([suspended_name, remaining_time - extra_time])
                    break
                    
        # 현재 과제를 완료하지 못하는 경우
        else:
            remaining = current_duration - (next_start - current_start)
            stack.append([current_name, remaining])
    
    # 마지막 과제 추가
    answer.append(tasks[-1][0])
    
    # 중단된 과제들을 순서대로 처리
    while stack:
        answer.append(stack.pop()[0])
    
    return answer

File: 엄마호랑이반/10주차 문제/장원석/과제진행하기.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Stack;

public class 과제진행하기 {

	public static void main(String[] args) {
		System.out.println(Arrays.toString(solution(new String[][] { { "korean", "11:40", "30" },
				{ "english", "12:10", "20" }, { "math", "12:30", "40" } })));
		System.out.println(Arrays.toString(solution(new String[][] { { "science", "12:40", "50" },
				{ "music", "12:20", "40" }, { "history", "14:00", "30" }, { "computer", "12:30", "100" } })));
		System.out.println(Arrays.toString(solution(
				new String[][] { { "aaa", "12:00", "20" }, { "bbb", "12:10", "30" }, { "ccc", "12:40", "10" } })));
	}

	static class Node {
		String name;
		int time;

		public Node(String name, int time) {
			this.name = name;
			this.time = time;
		}

		@Override
		public String toString() {
			return name + " " + time;
		}
	}

	private static String[] solution(String[][] plans) {
		int l = plans.length;
		List<String> answerList = new ArrayList<>();
		Stack<Node> stack = new Stack<>();

		Arrays.sort(plans, (p1, p2) -> {
			return p1[1].compareTo(p2[1]);
		});

		for (int i = 0; i < l - 1; i++) {
			String currentName = plans[i][0];
			int currentPlayTime = Integer.parseInt(plans[i][2]);

			String[] currentTimeStr = plans[i][1].split(":");
			int currentStart = Integer.parseInt(currentTimeStr[0]) * 60 + Integer.parseInt(currentTimeStr[1]);

			String[] nextTimeStr = plans[i + 1][1].split(":");
			int nextStart = Integer.parseInt(nextTimeStr[0]) * 60 + Integer.parseInt(nextTimeStr[1]);

			// 현재 과제와 다음 과제 시작 시간의 차이
			int gap = nextStart - currentStart;

			if (gap >= currentPlayTime) {
				// 현재 과제를 완료할 수 있는 경우
				answerList.add(currentName);

				// 남은 시간동안 스택의 과제 처리
				int remainTime = gap - currentPlayTime;

				while (remainTime > 0 && !stack.isEmpty()) {
					Node node = stack.peek();

					if (remainTime >= node.time) {
						remainTime -= node.time;
						answerList.add(stack.pop().name);
					} else {
						node.time -= remainTime;
						break;
					}
				}
			} else {
				// 현재 과제를 완료할 수 없는 경우
				stack.push(new Node(currentName, currentPlayTime - gap));
			}
		}

		// 마지막 과제 처리
		answerList.add(plans[l - 1][0]);

		// 스택에 남은 과제들 처리
		while (!stack.isEmpty()) {
			answerList.add(stack.pop().name);
		}

		return answerList.toArray(new String[0]);

	}

}
